### PR TITLE
Exposing Webternet functionality as natives

### PR DESCRIPTION
**Pull request by nefarius**
This patch exposes the HTTP logic of the Webternet extension as SourcePawn natives. 
See `plugins/include/webternet.inc` for a list of well commented natives. 

A sample plugin demonstrating the usage can be found [here](http://www.xup.to/dl,90855285/webtest.sp/).

### CODE REVIEW

**Comment by KyleSanderson** [comment]
@nefarius any update on this?

**Comment by nefarius** [comment]
@KyleSanderson sorry, not really, being very busy with my own projects =/

**Comment by ErikMinekus** [comment]
@nefarius It's unclear to me which bugs there are, care to elaborate?

Also, if you can merge master into your branch, I can make a PR for methodmaps.

**Comment by nefarius** [comment]
@ErikMinekus as far as i can recall there is a problem with in progress downloads when a plugin gets unloaded. This was also my first attempt to extend SourceMod so my coding style back then wasn't the best.

**Comment by ErikMinekus** [comment]
I was finally able to compile and run this extension on OS X, so I can take a look at the issues with my limited C++ knowledge.

One thing I'm wondering though, GetBuffer() and GetSize() are part of IBaseDownloader, but I think these are only needed for MemoryDownloader. FileDownloader shouldn't expose the actual file contents. Anyone have an opinion on this?

File: extensions/curl/extension.cpp [hunk 1]
```diff
@@ -86,3 +127,695 @@ const char *CurlExt::GetExtensionDateString()
...
+	// Use file-based download handler
+	FileDownloader *downloader = new FileDownloader(localPath);
+
+	// This should never happen but, oh well...
+	if (!downloader)
+	{
+		return pCtx->ThrowNativeError("Could not create downloader");
+	}
+
+	return g_pHandleSys->CreateHandle(g_DownloadHandle,
```
Comment (KyleSanderson) [inline extensions/curl/extension.cpp]: Leak on Handle Creation Failure x3.

File: extensions/curl/extension.cpp [hunk 2]
```diff
@@ -86,3 +127,695 @@ const char *CurlExt::GetExtensionDateString()
...
+static cell_t HTTP_CreateSession(IPluginContext *pCtx, const cell_t *params)
+{
+	IWebTransfer *x = g_webternet.CreateSession();
+
+	if (!x)
+	{
+		return pCtx->ThrowNativeError("Could not create session");
+	}
+
+	return g_pHandleSys->CreateHandle(g_SessionHandle,
```
Comment (KyleSanderson) [inline extensions/curl/extension.cpp]: ditto on failure leak.

File: extensions/curl/extension.cpp [hunk 3]
```diff
@@ -86,3 +127,695 @@ const char *CurlExt::GetExtensionDateString()
...
+	{
+		return pCtx->ThrowNativeError("HTTP web form data not found\n");
+	}
+
+	char *url;
+	// 4th param: target URL
+	pCtx->LocalToString(params[4], &url);
+
+	HTTPRequestCompletedContextPack contextPack;
+	contextPack.pCallbackFunction = new HTTPRequestCompletedContextFunction;
```
Comment (KyleSanderson) [inline extensions/curl/extension.cpp]: Memory leak, this is never deleted from the union declaration.

File: extensions/curl/extension.cpp [hunk 4]
```diff
@@ -86,3 +127,695 @@ const char *CurlExt::GetExtensionDateString()
...
+	m_pWorker = threader->MakeWorker(NULL, true);
+	m_pWorker->Start();
+}
+
+void HTTPSessionManager::HTTPAsyncRequestHandler::RunThread(IThreadHandle *pHandle)
+{
+	HandleError herr;
+	IPlugin *parent = plsys->PluginFromHandle(request.plugin, &herr);
+
+	if (parent == NULL || herr == HandleError_Freed)
```
Comment (KyleSanderson) [inline extensions/curl/extension.cpp]: SM is not thread-safe.

**Comment by psychonic** [comment]
I rebased this on master and pushed to a branch on our repository so that the team can more easily hack on it. I'll put up a new PR for further feedback once some of the current issues are addressed.

**Comment by ErikMinekus** [comment]
Maybe the last 5 commits of nefarius/sourcemod#1 are of interest, I mostly attempted to add methodmaps and did some code cleanup.

**Comment by psychonic** [comment]
@ErikMinekus Thanks!

I was able to cherry pick the first three. The last two weren't able to cleanly apply after the changes I had already made.

**Comment by ErikMinekus** [comment]
The last two are just moving the natives to natives.cpp, should be easy to redo manually.

**Comment by psychonic** [comment]
> The last two are just moving the natives to natives.cpp, should be easy to redo manually.

Fair point. That's been done now.

**Comment by Bara20** [comment]
Comes with ~ SourceMod 1.9 (hopefully) or later?

**Comment by nefarius** [comment]
Genuinely sorry this is such a mess :(

**Comment by KyleSanderson** [comment]
Hey man, you did most of the work. Now it's just fixing it up before merging, no need to apologise.

**Comment by jonatan1024** [comment]
Hi, what is the status of this PR? How much work is left to be done? Are there still plans to include these natives?

**Comment by psychonic** [comment]
Still abandoned and waiting for anyone to pick it back up, as the details above show.

**Comment by nefarius** [comment]
Oh Jesus, who'd have thought this is still in the queue 😲 Quite a few years have passed, I had time to improve my C++ skill tremendously 🤣 maybe I'll give a re-write a shot.

**Comment by KyleSanderson** [comment]
@asherkin so, I'm still inclined to close this (as I was 4 years ago). Are we still looking to keep this one open?

**Pull request closed without merging**